[libpas] PGM de-duplicate code across operating systems setting upper and lower guard pages

stwrt opened the pull request:
#### 3ae8c742d35327b4602092eac66da6ff0a59e6a8
<pre>
[libpas] PGM de-duplicate code across operating systems setting upper and lower guard pages
<a href="https://bugs.webkit.org/show_bug.cgi?id=293697">https://bugs.webkit.org/show_bug.cgi?id=293697</a>
<a href="https://rdar.apple.com/problem/152178338">rdar://problem/152178338</a>

Reviewed by Yusuke Suzuki and Mark Lam.

Windows support was added to libpas and this added duplicate code for setting the lower and upper guard pages. Let&apos;s share that code across platforms.
Only the mprotect/advice and VirtualAlloc calls need to be different.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):

Canonical link: <a href="https://commits.webkit.org/295536@main">https://commits.webkit.org/295536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4898f209567421aa423f9984aee005027774e6b6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25050 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110546 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55997 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33593 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80009 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108344 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19882 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95078 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60316 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13156 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55390 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98006 "Built successfully and passed tests") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89358 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113205 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103956 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32494 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23950 "") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89084 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91296 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88724 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22638 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33616 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11410 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27970 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32418 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/128260 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32191 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35063 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35536 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33764 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -175,27 +175,23 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+     */
     PAS_ASSERT(VirtualFree((void *) lower_guard, lower_guard_size, MEM_DECOMMIT));
     PAS_ASSERT(VirtualFree((void *) upper_guard, upper_guard_size, MEM_DECOMMIT));
 #else
```
We have contemplated making PAS_ASSERT behave like a typical ASSERT (i.e. no-op when not ASSERT_ENABLED), so while you're here you might as well split these up into two statements it like lines 199-200.

stwrt replied:
Fixed. Thanks for the heads up.

dhecht reviewed (commented):
LGTM (non-binding)

MenloDorian reviewed (approved):
r=me

stwrt committed:
[libpas] PGM de-duplicate code across operating systems setting upper and lower guard pages
https://bugs.webkit.org/show_bug.cgi?id=293697
rdar://problem/152178338

Reviewed by Yusuke Suzuki and Mark Lam.

Windows support was added to libpas and this added duplicate code for setting the lower and upper guard pages. Let's share that code across platforms.
Only the mprotect/advice and VirtualAlloc calls need to be different.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):

Canonical link: https://commits.webkit.org/295536@main
